File: ArbolesBinarios/binaryTree.py
class Node:
	def __init__(self, data):
		self.value = data
		self.left = None
		self.right = None

class binaryTree:
	def __init__(self):
		self.root = None

	def isEmpty(self):
		if self.root is None: 
			return True
		else: 
			return False
	
	def insert(self, data):
		if self.isEmpty():
			self.root = Node(data)
		elif self.find(self.root, data) is not None:
			print("The element {} already exists".format(data))
		else:
			self.add(self.root, data)
	def add(self, node, data):
		if data < node.value:
			if node.left is not None:
				self.add(node.left, data)
			else: node.left = Node(data)
		else:
			if node.right is not None:
				self.add(node.right, data)
			else:
				node.right = Node(data)

	def delete(self, data):
		if self.isEmpty():
			print("Invalid action. The tree is empty.")
		else:
			ref = self.find(self.root, data)
			if ref is None:
				print(f"The node {data} does not exists")
			elif ref.left is None and ref.right is None:		#Case 1. The node has no childs
				parent = self.findParent(self.root, data)
				ref = None
				if data < parent.value: parent.left = None
				else: parent.right = None
			elif ref.left is not None and ref.right is None:	#Case 2: The node has one child: left
				parent = self.findParent(self.root, data)
				if data < parent.value: parent.left = ref.left
				else: parent.right = ref.left
				ref = None
			elif ref.left is None and ref.right is not None:	#Case 2: The node has one child: right
				parent = self.findParent(self.root, data)
				if data < parent.value: parent.left = ref.right
				else: parent.right = ref.right
				ref = None
			elif ref.left is not None and ref.right is not None:#Case 3: The node has two childs
				minimum = self.findMinimumRight(ref.right)
				flag = minimum
				self.delete(minimum.value)
				ref.value = flag.value

	def find(self, node, data): #Returns the node if exists
		if node == None:
			return None
		elif data == node.value:
			return node
		elif data < node.value:
			return self.find(node.left, data)
		else:
			return self.find(node.right, data)

	def findParent(self, node, data): #Returns the node's parent
		if data < node.value: #Goes to the left tree
			if node.left.value == data:
				return node
			return self.findParent(node.left, data)
		else:   #Goes to the right node
			if data == node.right.value:
				return node
			return self.findParent(node.right, data)

	def findMinimumRight(self, current_node):
		while current_node.left is not None:
			current_node = current_node.left
		return current_node

	def search(self, data): #Just a boolean to determine if exists
		if self.isEmpty(): print("The tree is empty. Insert something :)")
		else: self._search(self.root, data)
	def _search(self, node, data):
		if data == node.value:
			print("Element {} exists.".format(data))
		elif data < node.value and node.left is not None:
			self._search(node.left, data)
		elif data > node.value and node.right is not None:
			self._search(node.right, data)
		else:
			print("Element {} does not exists.".format(data))

	def show(self):
		if self.isEmpty():
			print("The tree is empty. Insert something", end="\n")
		else:
			self.inorder() #Inorder displays items in ascending order
			print()

	def inorder(self):
		if self.isEmpty():
			print("The tree is empty. Insert something", end="\n")
		else:
			self._inorder(self.root)
			print()
	def _inorder(self, node): #Displays items as left tree, parent, right tree
		if node is not None:
			self._inorder(node.left)
			print("{} ->".format(node.value), end=" ")
			self._inorder(node.right)
	
	def preorder(self):
		if self.isEmpty():
			print("The tree is empty. Insert something", end="\n")
		else:
			self._preorder(self.root)
			print()
	def _preorder(self, node): #Display items as parent, left tree, right tree
		if node is not None:
			print("{} ->".format(node.value), end=" ")
			self._preorder(node.left)
			self._preorder(node.right)

	def postorder(self):
		if self.isEmpty():
			print("The tree is empty. Insert something", end="\n")
		else:
			self._postorder(self.root)
			print()
	def _postorder(self, node): #Displays items as left tree, rith tree, parent
		if node is not None:
			self._postorder(node.left)
			self._postorder(node.right)
			print("{} ->".format(node.value), end=" ")

	

if __name__ == "__main__":
	tree = binaryTree()

	#Put here a menu or the options you would like your list to do.
	#Functions are insert, delete, inorder, preorder, postorder, search and show

	tree.insert(8)
	tree.insert(4)
	tree.insert(12)
	tree.insert(2)
	tree.insert(6)
	tree.insert(10)
	tree.insert(14)
	tree.insert(1)
	tree.insert(3)
	tree.insert(5)
	tree.insert(7)
	tree.insert(9)
	tree.insert(11)
	tree.insert(13)
	tree.insert(15)
	tree.inorder()
	tree.delete(8)
	tree.inorder()
	tree.postorder()
	tree.preorder()

File: ArbolesBinarios/binaryTree.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace binaryTree
{
    class Node
    {
        public int value;
        public Node left, right;

        public Node(int data)
        {
            value = data;
        }
    }

    public class binaryTree
    {
        private Node root;

        public binaryTree()
        {
            root = null;
        }

        bool isEmpty()
        {
            if (root == null) return true;
            else return false;
        }
        void insert(int data)
        {
            if (isEmpty()) root = new Node(data);
            else if (find(root, data) != null)
                Console.WriteLine("The element " + data + " already exists.");
        else insert(root, data);
        }
        void insert(Node node, int data)
        {
            if (data < node.value)
            {
                if (node.left != null) insert(node.left, data);
                else
                {
                    node.left = new Node(data);
                }
            }
            else
            {
                if (node.right != null) this.insert(node.right, data);
                else node.right = new Node(data);
            }
        }

        void remove(int data)
        {
            if (isEmpty()) Console.WriteLine("Invalid action. The tree is empty.");
        else
            {
                Node reference = find(root, data);
                if (reference == null) Console.WriteLine("The node " + data + " does not exists.");
            else if (reference.left == null && reference.right == null)
                {
                    Node parent = findParent(root, data);
                    reference = null;
                    if (data < parent.value) parent.left = null;
                    else parent.right = null;
                }
                else if (reference.left != null && reference.right == null)
                {
                    Node parent = findParent(root, data);
                    if (data < parent.value) parent.left = reference.left;
                    else
                    {
                        parent.right = reference.left;
                        reference = null;
                    }
                }
                else if (reference.left == null && reference.right != null)
                {
                    Node parent = this.findParent(root, data);
                    if (data < parent.value) parent.left = reference.right;
                    else
                    {
                        parent.right = reference.right;
                        reference = null;
                    }
                }
                else if (reference.left != null && reference.left != null)
                {
                    Node minimum = findMinimumRight(reference.right);
                    Node flag = minimum;
                    remove(minimum.value);
                    reference.value = flag.value;
                }
            }
        }

        Node find(Node node, int data)
        {
            if (node == null) return null;
            else if (data == node.value) return node;
            else if (data < node.value) return find(node.left, data);
            else return find(node.right, data);
        }

        Node findParent(Node node, int data)
        {
            if (data < node.value)
            {
                if (node.left.value == data) return node;
                return findParent(node.left, data);
            }
            else
            {
                if (data == node.right.value) return node;
                return findParent(node.right, data);
            }
        }

        Node findMinimumRight(Node node)
        {
            while (node.left != null)
                node = node.left;
            return node;
        }

        void search(int data)
        {
            if (isEmpty()) Console.WriteLine("The tree is empty. Insert something :)");
            else search(root, data);
        }
        void search(Node node, int data)
        {
            if (data == node.value) Console.WriteLine("Element " + data + " exists.");
            else if (data < node.value && node.left != null)
                    search(node.left, data);
                else if (data > node.value && node.right != null)
                    search(node.right, data);
                else Console.WriteLine("Element " + data + " does not exists.");
        }

        void show()
        {
            if (isEmpty()) Console.WriteLine("The tree is empty. Insert something");
            else inorder();
        }
        void inorder()              //Displays items as left tree, parent, rigth tree
        {
            if (isEmpty()) Console.WriteLine("The tree is empty. Insert something");
            else
                {
                    inorder(root);
                    Console.WriteLine();
                }
        }
        void inorder(Node node)    //Use Polymorphism to allow recursive methods
        {
            if (node != null)
            {
                inorder(node.left);
                Console.Write(node.value + " -> ");
                inorder(node.right);
            }
        }

        void preorder()              //Displays items as parent, left tree, rigth tree
        {
            if (isEmpty()) Console.WriteLine("The tree is empty. Insert something");
        else
            {
                preorder(root);
                Console.WriteLine();
            }
        }
        void preorder(Node node)    //Use Polymorphism to allow recursive methods
        {
            if (node != null)
            {
                Console.Write(node.value + " -> ");
                preorder(node.left);
                preorder(node.right);
            }
        }

        void postorder()              //Displays items as left tree, rigth tree, parent
        {
            if (isEmpty()) Console.WriteLine("The tree is empty. Insert something");
            else
                {
                    postorder(root);
                    Console.WriteLine();
                }
            }
        void postorder(Node node)    //Use Polymorphism to allow recursive methods
        {
            if (node != null)
            {
                postorder(node.left);
                postorder(node.right);
                Console.Write(node.value + " -> ");
            }
        }
   
        static void Main(string[] args)
        {
            binaryTree tree = new binaryTree();

            //Put here a menu or the options you would like your list to do.
            //Functions are insert, remove, inorder, preorder, postorder, search and show

            tree.insert(8);
            tree.insert(4);
            tree.insert(12);
            tree.insert(2);
            tree.insert(6);
            tree.insert(10);
            tree.insert(14);
            tree.insert(1);
            tree.insert(3);
            tree.insert(5);
            tree.insert(7);
            tree.insert(9);
            tree.insert(11);
            tree.insert(13);
            tree.insert(15);
            tree.show();
            tree.remove(1);
            tree.remove(3);
            tree.remove(5);
            tree.remove(7);
            tree.remove(13);
            tree.remove(15);
            tree.remove(9);
            tree.remove(11);
            tree.inorder();
            tree.preorder();
            tree.postorder();

            Console.ReadLine();
        }
    }
}
